rllib/core/rl_trainer/tf/tf_rl_trainer.py

```diff
@@ -0,0 +1,150 @@
+import numpy as np
+from typing import Any, Mapping, Union
+from ray.rllib.core.rl_trainer.rl_trainer import RLTrainer
+from ray.rllib.core.rl_module.rl_module import RLModule, ModuleID
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.framework import try_import_tf
+from ray.rllib.utils.typing import TensorType
+from ray.rllib.utils.nested_dict import NestedDict
+from ray.rllib.utils.numpy import convert_to_numpy
+
+tf1, tf, tfv = try_import_tf()
+tf1.enable_eager_execution()
+
+
+class TfRLTrainer(RLTrainer):
+    """Base class for RLlib TF algorithm trainers."""
+
+    def remove_module(self, module_id: ModuleID):
+        self._module.remove_module(module_id)
+
+    def add_module(self, module_id: ModuleID, module: RLModule):
+        self._module.add_module(module_id, module)
+
+    @tf.function
+    def _do_update_fn(self, batch: SampleBatch) -> Mapping[str, Any]:
+        with tf.GradientTape() as tape:
+            fwd_out = self._module.forward_train(batch)
+            loss = self._rl_optimizer.compute_loss(batch, fwd_out)
+            if isinstance(loss, tf.Tensor):
+                loss = {"total_loss": loss}
+        gradients = self.compute_gradients(loss, tape)
+        self.apply_gradients(gradients)
+        return {"loss": loss, "fwd_out": fwd_out, "post_processed_gradients": gradients}
+
+    def update(self, batch: SampleBatch) -> Mapping[str, Any]:
+        """Perform an update on this Trainer.
+
+        Args:
+            batch: A batch of data.
+
+        Returns:
+            A dictionary of results.
+        """
+        # TODO(sven): This is a hack to get around the fact that
+        # SampleBatch.count becomes 0 after decorating the function with
+        # tf.function. This messes with input spec checking. Other fields of
+        # the sample batch are possibly modified by tf.function which may lead
+        # to unwanted consequences. We'll need to further investigate this.
+        batch = NestedDict(batch.policy_batches)
+        for key, value in batch.items():
+            batch[key] = tf.convert_to_tensor(value, dtype=tf.float32)
+        if self.distributed:
+            infos = self.strategy.run(self._do_update_fn, args=(batch,))
```

Comment:
`do_distrubuted_update()` can be defined at the level of `RLTrainer` so that the logic inside this `.update()` function can be pulled out from here and moved to RLTrainer to make a deeper class.